src/textual/_time_to_completion.py

```diff
@@ -0,0 +1,301 @@
+"""Code to help calculate the estimated time to completion of some process."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from time import monotonic
+
+from rich.repr import Result
+from typing_extensions import Self
+
+
+@dataclass
+class Sample:
+    """A sample."""
+
+    value: float
+    """The value of the sample."""
+
+    moment: float
+    """The moment when the sample was taken."""
+
+
+class Samples:
+    """A deque-ish-like object that holds samples.
+
+    The window for the samples is based on time, with a maximum sample cap
+    within that window to ensure memory use doesn't run away. Also, for
+    times when the sample window starts to run dry (drops to just the one
+    available sample in the window) the [last-dropped
+    sample][textual._time_to_completion.Samples.last_dropped] is held on to
+    and made available.
+    """
+
+    def __init__(self, time_window_size: float, max_samples: int) -> None:
+        """Initialise the samples object.
+
+        Args:
+            time_window_size: The window of time to keep samples for.
+            max_samples: The maximum number of samples to keep.
+        """
+        self._time_window_size = time_window_size
+        """The maximum amount of time to keep the samples for."""
+        self._max_samples = max_samples
+        """The maximum number of samples to keep."""
+        self._samples: list[Sample] = []
+        """The samples."""
+        self._last_dropped: Sample | None = None
+        """Keep track of the last-dropped sample."""
+
+    def _prune(self) -> Self:
+        """Prune the samples.
+
+        Returns:
+            Self.
+        """
+        if samples := self._samples:
+            # Trim off any "too old" samples.
+            oldest_time = samples[-1].moment - self._time_window_size
+            for position, sample in enumerate(samples):
+                if sample.moment > oldest_time:
+                    # We keep the "last dropped" sample around for any sort
+                    # of fallback position.
+                    self._last_dropped = samples[position - 1]
+                    self._samples = samples[position:]
+                    break
+            # Ensure that we don't run up too many samples.
+            self._samples = self._samples[-self._max_samples :]
+        return self
+
+    def append(self, sample: Sample) -> Self:
+        """Add a sample to the samples.
+
+        Args:
+            sample: The sample to add.
+
+        Returns:
+            Self.
+        """
+        self._samples.append(sample)
+        return self._prune()
```

Comment:
Do we need to call prune on every sample?

Could we defer pruning to when we need the samples? i.e. if there are 1000 samples between refreshes, could we prune just once?

Reply:
This could go either way; there could be 1,000 "refreshes" between samples too. It feels to me like, on balance, code that is showing an estimated time is more likely to get new estimates than it is to record new samples.